fars/api/views.py

```diff
@@ -24,6 +27,65 @@ class BookingsList(viewsets.ViewSetMixin, generics.ListAPIView):
     filter_backends = (filters.DjangoFilterBackend,)
     filter_class = BookingFilter
 
+    def filter_on_bookable(self, bookings, request):
+        bookable = request.query_params.get('bookable')
+        if bookable:
+            return bookings.filter(bookable__id_str=bookable), True
+        return bookings, False
+
+    def serialize_one(self, booking):
+        # If the booking is None, the serializer will create a null object, which we don't want
+        return Response(self.get_serializer_class()(booking).data if booking else None)
+
+    def serialize_many(self, bookings):
+        # If the booking is None, the serializer will create a null object, which we don't want
+        return Response(self.get_serializer_class()(bookings, many=True).data)
+
+    def filter_on_bookable_and_serialize_many(self, bookings, request):
+        bookings, _ = self.filter_on_bookable(bookings, request)
+        return self.serialize_many(bookings)
+
+    @action(methods=['get'], detail=False)
+    def future(self, request):
+        bookings = self.get_queryset().filter(start__gte=datetime.datetime.now())
+        return self.filter_on_bookable_and_serialize_many(bookings, request)
+
+    @action(methods=['get'], detail=False)
+    def past(self, request):
+        bookings = self.get_queryset().filter(end__lte=datetime.datetime.now())
+        return self.filter_on_bookable_and_serialize_many(bookings, request)
+
+    @action(methods=['get'], detail=False)
+    def next(self, request):
```

Comment:
This isn't great design, typically you would have a booking in the request and find the next booking following that booking. This way you can get the "next" booking from the reference of any booking, not just current booking. Even better would be to just use the `future` and `past` endpoints and have the frontend pick the first from the response.

Reply:
Good point, that could be generalized.

Reply:
And picking the first response from a list is only possible if the order is known.

Reply:
But the responses seem to be ordered by start time (and not for example by id), but this is not completely obvious.

Reply:
You can let the client pass an ordering parameter or enforce ordering by time (you should not trust the ordering it if it's not explicit)